src/QMCWaveFunctions/detail/CUDA/cuBLAS_LU.cu

```diff
@@ -0,0 +1,212 @@
+  diag.x              = lu_diag.x * pivot_factor;
+  diag.y              = lu_diag.y * pivot_factor;
+  cuDoubleComplex log_value;
+  log_value.x = log(sqrt(diag.x * diag.x + diag.y * diag.y));
+  log_value.y = atan2(diag.y, diag.x);
+  return log_value;
+}
+
+/** computes the log determinant using the output of LU factorization.
+   */
+template<int COLBS, typename T>
+__global__ void computeLogDet_kernel(const int n,
+                                     const int lda,
+                                     T** mat_lus,
+                                     const int* const pivots,
+                                     cuDoubleComplex* logdets)
+{
+  const int iw                = blockIdx.x;
+  const int block_num         = blockIdx.y;
+  const T* __restrict__ lu_iw = mat_lus[iw];
+  __shared__ cuDoubleComplex logdet_vals[COLBS];
+  T lu_diag;
+  int n_index = threadIdx.x + block_num * COLBS;
+  if (n_index < n)
+  {
+    lu_diag                  = *(lu_iw + n_index * lda + n_index);
+    logdet_vals[threadIdx.x] = complexDetLog(lu_diag, n_index, pivots);
+  }
+  if (threadIdx.x == 0 && blockIdx.y == 0)
+    logdets[iw] = {0.0, 0.0};
+  // insure that when we reduce logdet_vals all the threads in the block are done.
+  __syncthreads();
+  if (threadIdx.x == 0)
+  {
+    __shared__ cuDoubleComplex block_sum_log_det;
+    block_sum_log_det = {0.0, 0.0};
+    for (int iv = 0; iv < COLBS; ++iv)
+    {
+      block_sum_log_det.x += logdet_vals[iv].x;
+      block_sum_log_det.y += logdet_vals[iv].y;
+    }
+    // No atomicAdd in cuda 10 for cuComplex/cuDoubleComplex
+    atomicAdd((double*)&(logdets[iw].x), block_sum_log_det.x);
+    atomicAdd((double*)&(logdets[iw].y), block_sum_log_det.y);
+  }
+}
+
+/** Calculates logdets using LU_diags and pivots
+ *  \param[in] LU_mat - the LU output from cublasXgetrfBatched
+ *  \param[out] LU_diags - the LU_diags from the LU
+ *  \param[in] batch_size - no a big deal here.
+ */
+template<typename TMAT, typename T>
+cudaError_t computeLogDet_batched_impl(cudaStream_t& hstream,
+                                       const int n,
+                                       const int lda,
+                                       TMAT** LU_mat,
+                                       const int* pivots,
+                                       T* logdets,
+                                       const int batch_size)
+{
+  // Perhaps this should throw an exception. I can think of no good reason it should ever happen other than
+  // developer error.
+  if (batch_size == 0 || n == 0)
+    return cudaSuccess;
+
+  const int COLBS          = 256;
+  const int num_col_blocks = (n + COLBS - 1) / COLBS;
+  dim3 dimBlock(COLBS);
+  dim3 dimGrid(batch_size, num_col_blocks);
+  computeLogDet_kernel<COLBS>
+      <<<dimGrid, dimBlock, 0, hstream>>>(n, lda, CUDATYPECAST(LU_mat), pivots, CUDATYPECAST(logdets));
+  return cudaPeekAtLastError();
+}
+
+template<typename T>
+void computeLogDet_batched(cudaStream_t& hstream,
```

Comment:
Is this function actually used in real code?
I thought computeInverseAndDetLog_batched is the way to go.

Reply:
no this is for testing only.

Reply:
This likely will be an outside API call in a future PR.